[DOCS] first draft for Data Quality: Schema tech doc

Quantisan opened the pull request:
NOT READY FOR MERGE. This PR is a request for a review of our first draft to ensure we're on the right path in terms of content style and tone. The mechanics of the doc (e.g. code CI and styling syntax) are not in place yet. 


- [ ] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x ] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [ ] Code is linted - run `invoke lint` (uses `ruff format` + `ruff check`)
- [ ] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

**Data quality dimension doc template:**
Here is a checklist based on the draft article template outline that we discussed. Thanks!
- [ ] 1. An explanation of the data quality issue/dimension
- [ ] 2. A short description of the real world implications of this data quality issue (e.g. consequences to particular stakeholders, business objectives)
- [ ] 3. An exploration of Expectations that are relevant to addressing the data quality issue at hand. 
- [ ] 4. A GX recipe that demonstrates how to identify this data quality issue in a dataset, using Expectations:
  * A small sample dataset (provided as CSV, Snowflake, and Postgres format)
  * A functional code example of using GX to identify the DQ issue using the sample data
- [ ] 5. A "Next steps" concluding section that describes, as applicable:
  * Complex considerations for identifying and monitoring the data quality issue (to demonstrate that we've thought deeply about how it can and should be validated)
  * Related GX resources

Quantisan committed:
first draft for schema tech doc

rachhouse commented:
@Quantisan Hey Paul, thanks for creating this first draft as a PR! Here are my overall comments:

* First, I added a draft template outline for these docs in our shared slack channel. Once we've had a chance to discuss the template tomorrow, I think the next step will be to reshape and augment this draft into that format.

* Including relevant Expectations to the covered data quality topic is great - I'd love to see more technically nuanced content than just a restating of the Expectation name. For instance, why might a user opt to use `expect_table_columns_to_match_ordered_list` over `expect_table_columns_to_match_set`? We want to go deeper than just "here's our Expectation" so that users can appreciate that, as a data quality company, we've really considered (and can elaborate and demonstrate) how we're non-superficially addressing a data quality issue with GX. 

* We can assume for the purposes of these data quality dimension/issue docs that users are familiar with the basic components of a GX workflow (defining Expectations, creating an Expectation Suite, running a Checkpoint, reviewing Validations). We want to focus on GX technical specifics that are relevant and unique to the data quality topic at hand - which I suspect will mostly center on how to effectively employ the associated Expectations. I think we can use lines of code in our examples that implement the common parts of a GX workflow (steps that will occur in all these documentation examples, like creating Checkpoints, running Checkpoints, saving Expectation Suites) without needing to explain each line.

* Do you have a sample dataset in mind or defined for this doc?

I think your draft is a good start - and arguably this first one will probably take a bit more back and forth since we're hashing out the pattern that we'll use for the subsequent DQ docs.

Quantisan committed:
incorporate @rachhouse's feedback, particular point (2)

https://github.com/great-expectations/great_expectations/pull/10022#issuecomment-2161590238

Quantisan commented:
> * Do you have a sample dataset in mind or defined for this doc?

Yes, we created a dataset generator for the financial compliance story that we can use for this whole DQ series (both blog and articles).

Quantisan commented:
> * Including relevant Expectations to the covered data quality topic is great - I'd love to see more technically nuanced content than just a restating of the Expectation name. For instance, why might a user opt to use `expect_table_columns_to_match_ordered_list` over `expect_table_columns_to_match_set`? We want to go deeper than just "here's our Expectation" so that users can appreciate that, as a data quality company, we've really considered (and can elaborate and demonstrate) how we're non-superficially addressing a data quality issue with GX.

thanks for your feedback, @rachhouse. We've worked in your suggestions into this latest version, [87b8485](https://github.com/great-expectations/great_expectations/pull/10022/commits/87b84856671edfd32d326627784a235265180569). Have another look when you have a chance?

Quantisan committed:
update #1

Quantisan committed:
incorporate some notes from Data Quality Issues doc

Quantisan committed:
update intro and conclusion, add a steps list

Quantisan committed:
copy edit the list of schema Expectations

Quantisan committed:
fix a subtitle

Quantisan committed:
stub to prepare sample data

Quantisan committed:
add sample data step to the schema tutorial

rachhouse commented on docs/docusaurus/docs/core/data_quality/01_schema.md:
````diff
@@ -0,0 +1,204 @@
+errors due to missing fields.
+
+```python
+dataset.expect_column_to_exist("sender_account_number")
````
What do you think about introducing the sample data used throughout the documentation article before referring to it in Expectation code snippets? Otherwise, users lack understanding of the data columns and data are before seeing them used in examples.

Quantisan replied:
We experimented with this a bit and settled on showing a data preview in a HTML table. Our intention is to dive into the Expectations quickly to keep the doc engaging in the beginning.

rachhouse commented on docs/docusaurus/docs/core/data_quality/01_schema.md:
```diff
@@ -0,0 +1,204 @@
+**Use Case**: Ideal during data ingestion or integration of multiple data sources to ensure that
+essential fields are present before proceeding with downstream processing.
+
+**Pro Tip**: Implement this Expectation early in your data pipeline to catch missing columns as soon
```
I like the way you've included Pro Tips/Strategic Advice/Best Practices! I also think it would be great if we could keep these bonus points consistently named, so that there's a consistent pattern to the way we present any Expectation within and across data quality use cases. Maybe "GX Tip" or "GX Advice?" (I'm open to naming suggestions).

rachhouse commented on docs/docusaurus/docs/core/data_quality/01_schema.md:
```diff
@@ -0,0 +1,204 @@
+
+### Integrating Schema Validation
+
+Below is a tutorial that demonstrates schema validation of a Pandas DataFrame using GX.
```
Our goal for the use case documentation is to assume that the user/reader is familiar with the basic GX workflow mechanics (and we can always include a link to prerequisite reading if necessary). As-is, steps 2 and 4-7 will be the same across all data quality use cases - more boilerplate than insightful. In addition to the downsides of maintaining duplicated content, I think it might be easy for the user to miss what's actually relevant when it comes to the DQ issue at hand.

I think we can make the code recipes more impactful by focusing on what is unique to the data quality issue at hand. For example, potentially:
* providing two datasets and using the same Expectation Suite to check schema, and then explaining how the validation results differ for the datasets and why
* creating & running different Expectation Suites that use different parameters for the schema-related Expectations and showing how that impacts the validation results

(I'm also open to ideas if you have any thoughts!)

Quantisan replied:
We summarized the `Integrating Schema Validation` section and added an `Examples & Scenarios` section. Note that the code in the examples is currently pseudo code, but it should give you a good idea of what it'll look like.

Quantisan replied:
section is removed https://github.com/great-expectations/great_expectations/pull/10022#discussion_r1655206253

rachhouse commented on docs/docusaurus/docs/core/data_quality/01_schema.md:
```diff
@@ -0,0 +1,204 @@
+## Data Quality: Ensuring Schema Consistency with Great Expectations
+
+When working with data, you might face common challenges that necessitate a data quality platform.
```
I think your content for the opening paragraph is good, but I'd like to see the definition of data schema expressed more clearly and succinctly upfront vs. woven into the message of why it's important to validate. (We definitely want to keep the explanation of why it's important - but I'd like the content to distinguish the _what_ and the _why_ more distinctly for the user).

For each of the data quality dimensions that we create a use case for, I think it will be helpful to the user to begin with a concise understanding of "What is schema?"/"What is missingness?"/"What is <insert data quality dimension here>?"

Quantisan replied:
we were aiming for "show, don't tell", but including a concise explanation upfront would add clarity

rachhouse reviewed (commented):
@Quantisan Thanks for your latest updates! I've reviewed and provided some higher level comments on a few sections to inform the next revision.

Let me know if you have any questions, thanks!

Quantisan commented:
@cla-bot check

Quantisan committed:
move schema content into learn/

Quantisan committed:
expand sample dataset with more fields

Quantisan committed:
explain data schema in intro

fixes
https://github.com/great-expectations/great_expectations/pull/10022#discussion_r1648056397

Quantisan committed:
name all Tips as GX Tip

https://github.com/great-expectations/great_expectations/pull/10022#discussion_r1648028863

Quantisan committed:
show sample data upfront

https://github.com/great-expectations/great_expectations/pull/10022#discussion_r1648022627

Quantisan committed:
replace python data preview with html table

Quantisan committed:
remove instruction to boilerplate steps

Quantisan committed:
examples and scenarios

Quantisan committed:
add a Community Best Practices section

Quantisan committed:
note TODOs

Quantisan committed:
update conclusion to note need for a holistic DQ strategy

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,371 @@
 ---
 
-import InProgress from '../../../core/_core_components/_in_progress.md'
+**Data schema** refers to the structural blueprint of a dataset, encompassing elements such as column
```
Nice, I really like the new intro!

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,371 @@
+
+### Data Preview
+
+Here's a glimpse of the sample dataset we'll use to demonstrate schema validation:
```
Let's stick to consistent use of second person for the reader: https://developers.google.com/style/person

Here is an example rewording that avoids the `we'll` - feel free to modify as you see fit!

> Below is a sample of the dataset that is referenced by examples and explanations within this article. The sample data is modeled on financial transfers and contains columns such as `sender_account_number`, `recipient_fullname`, `transfer_amount`, and `transfer_date`.

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,371 @@
+
+### Key Schema Expectations
+
+Let's explore core Expectations for schema validation, delving into their practical applications and
```
First -> second person flag :)

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,371 @@
+Let's explore core Expectations for schema validation, delving into their practical applications and
+nuances:
+
+**1. `expect_column_to_exist`**:
```
I like the format you've opted for here - Expectation, snippet, use case, GX Tip. Rendered in Docusaurus it's difficult to visually parse the different Expectations presented. I'll take a look to see if Docusaurus has any visual elements or spacing we could use to tweak the rendering - definitely open to ideas if you have any.

I considered suggesting a table, but the main column of the docs page is fixed width and I don't think we have enough horizontal real estate to present the information in a readable way.

Quantisan replied:
![Screenshot 2024-06-27 at 09 27 31](https://github.com/great-expectations/great_expectations/assets/484035/c89d822a-7b72-41e3-a2fe-887d6c006c02)

Good catch! Let's put those tips in an admonition feature. For separating the Expectations, we could use subheadings or horizontal lines - I've added examples of both in #1 and #2 for you to review.

Since we're not sure about your team's preferred design style, I've provided a couple options so you can choose what works best.

Quantisan replied:
I moved all the headings up a level so now the Expectations are 3rd level subheadings that show up on the page TOC on the right-side bar.

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,371 @@
+**GX Tip**: Regularly review the allowed range as your dataset evolves, ensuring it aligns
+with business requirements and anticipates potential future expansion.
+
+### Integrating Schema Validation
```
I think we can consider this section (Integrating Schema Validation) content prerequisite knowledge for the data quality use case articles, since it's just the basics of a GX workflow. Since these steps will be the same for all DQ use articles, I'd vote to remove this section from the individual articles and move it to a separate GX workflow basics page, or just remove it. I can ask Rachel Reverie if there's any planned 1.0 content that will cover these steps in the way you have below.

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,371 @@
+import great_expectations as gx
+
+# Sample datasets
+data_1 = [
```
We can talk about this more in our meeting today, but the key decisions coming out of our meeting w/ Abe & Product yesterday were:
1. The code recipes must use a Postgres data source
2. Sample data can be supplied in CSV format - the articles will assume the user has the knowledge to populate their own Postgres data source with the sample data (we can provide a link for them)
3. The code recipes must use GX 1.0 (Core) code
4. ~~Each code recipe must be paired with a set of steps of how a user would achieve the same result using the GX Cloud UI~~
  * ~~There's a Docusaurus option for tabbed sections which I think could be a feasible way to present the code and UI variants, we can look at the syntax in our meeting~~

Rachel Reverie gave me a walkthrough of how we currently put code snippets in the docs (i.e. the recipes here) under test with sample data, so I can help you get set up with the right framework. We'll discuss/strategize when we chat today.

rachhouse replied:
**_Edited to add_:** Please ignore item 4 for now.

rachhouse replied:
For the previously struck item 4, I'll update the branch with the product `.gif`s once they are created.

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,371 @@
+
+**Insight**: The strict suite ensures that columns appear in the specified order, crucial in contexts where order matters for processing logic, while the relaxed suite allows flexibility but ensures all required columns are present.
+
+#### Step-by-Step Walkthrough: Managing Column Existence
```
I'm not completely clear on why this section follows a different pattern than the two previous examples. Is the intent to show the user a dissection of the validation results?

However, I do see the relevance of this section to the question you asked in Slack about error handling - working on tracking down more details.

Quantisan replied:
this dynamic pipeline example might be a bit too deep... My intention is on opening considerations for the overall lifecycle of a GX workflow, and what clever things you can do with it. Our previous examples stop at either success or failed validation. This section answers: I want to improve our data quality, now that we see success/fail on our data, now what?

Now that I speak out loud, this seems out of scope of this particular article. Same reason as the Integration list of steps, this can become boilerplate throughout the series.

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,371 @@
+
+**Insight**: Ensures that critical columns are present at key stages, allowing for early detection of schema inconsistencies, which minimizes downstream processing issues.
+
+#### Industry-Specific Scenarios: Healthcare Data Validation
```
I like having examples from different industries. Let's keep this for now, but I am going to check in with Abe to see if he has any opinions/requirements on length of these data quality use case articles. It's possible we might want to stick to 2-3 examples under one industry domain in the interests of a shorter read for the user.

Quantisan replied:
I agree the article is getting long. Actually...how about we remove this section? On second look, it doesn't add much value. since we're not healthcare data experts, too many industry examples could come across as contrived to those in the field (as you pointed out last week). let's keep it focused on our core points instead. what do you think?

rachhouse replied:
Yup, I'm good with removing this. Additionally, in conversation with Josh and Abe, both expressed interest in eventually having separate GX <> Industry use cases in the Learn docs. I think keeping with the single industry above and focusing on the DQ schema aspects (core points) is the way to go.

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,371 @@
+
+- [Great Expectations Official Documentation](https://docs.greatexpectations.io/)
+- [Common Schema Validation Pitfalls](#)
+- [Community Best Practices](#)
```
I was curious about the links here - are you thinking about moving the common pitfalls or best practices content to a different page?

Quantisan replied:
these links provide external resources on data quality to complement our content. we can curate some excellent materials here to enhance our coverage of this topic. by sharing these carefully selected links, we demonstrate our commitment to offering comprehensive data quality knowledge beyond just our GX tool

I added clarification in the TODO

Quantisan replied:
i did some research and found a couple promising links that seem relevant to schema validation. it's a bit tricky to discuss this without getting too deep into database implementation details, but i think these resources could be a good starting point. Otherwise, we can remove this section if it seems distracting.

[88105bd](https://github.com/great-expectations/great_expectations/pull/10022/commits/88105bdb8182dbb21891e9668274773fdeb678f1)

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,371 @@
+#### Common Pitfalls and How to Avoid Them
+
+- **Inconsistent Data Types**: Regularly monitor data ingestion points to ensure type consistency.
+- **Schema Evolution**: Establish a process for schema versioning and backward compatibility.
```
Love the pitfalls! Could you add a sentence or two to each that explicitly describes for the reader why they're pitfalls? (Since currently, it just notes how to avoid them).

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,371 @@
+
+To effectively handle these dimensions, consider integrating various Expectations to cover these broader data quality aspects. Regular validation, monitoring, and iterations are key to maintaining high standards.
+
+### The Path Forward
```
I appreciate this section, but I think we can end the article at the Conclusion, as we've covered some of these points in other sections, and it doesn't have unique relevance to schema.
* **Expand Beyond Schema** - covered nicely in the Conclusion
* **Implement Monitoring Solutions** - covered in Pitfalls
* **Foster a Data-centric Culture** and **Iterate and Evolve** - Regular review & update of Expectations is mentioned throughout the article. Both of these are important high level strategies for DQ, but could apply to any of the different DQ use cases, and we don't want to duplicate the same path forward across the use cases.

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,371 @@
+- [Common Schema Validation Pitfalls](#)
+- [Community Best Practices](#)
+
+### Conclusion
```
What do you think about renaming the _Conclusion_ to _The path forward_?

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,371 @@
 ---
 sidebar_label: 'Schema'
-title: 'Schema'
+title: 'Data Quality: Ensuring Schema Consistency with Great Expectations'
```
Overall style guide note - we use Sentence casing for our headings in the docs.

rachhouse reviewed (commented):
@Quantisan I really like your content and think it's evolving well. I expect the next thing we'll focus on are the code recipes - as noted in my comments, we now have more clarity on code and data specifics.

Quantisan committed:
address reader as second person

Quantisan committed:
try md admonition on GX Tip blocks

Quantisan committed:
add horizontal markers

Quantisan committed:
Expectations list as sub-headings

Quantisan committed:
remove integration steps

https://github.com/great-expectations/great_expectations/pull/10022#discussion_r1655206253

Quantisan committed:
note TODO in Community Resources

Quantisan committed:
use sentence casing for title

Quantisan committed:
remove mock healthcare example

Quantisan committed:
remove dynamic pipeline example

Quantisan committed:
expand on common pitfalls section to explain why

Quantisan committed:
replace The Path Forward section with existing Conslusion

https://github.com/great-expectations/great_expectations/pull/10022#discussion_r1655275467

Quantisan committed:
refactor all markdown headings one level up

Quantisan committed:
update list of future DQ topics

Quantisan committed:
add a pitfall on solely relying on schema validations

Quantisan committed:
Expectations items are now sub-headings instead of a list

Quantisan committed:
replace web links with a couple real external links

Quantisan committed:
all tip/info into admonitions

Quantisan committed:
test link to next DQ post

Quantisan committed:
fix heading to Resources section

Quantisan committed:
shorten post title

rachhouse commented:
@Quantisan I see the Netlify preview failed to build - I'm going to merge `develop` in to update the branch and retrigger the Netlify build.

rachhouse committed:
Merge branch 'develop' into docs/data-quality-schema-tech-doc

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,296 @@
+However, schema validation is just one aspect of a comprehensive data quality strategy. Achieving high-quality data requires a multifaceted approach involving:
+
+1. **Data Integrity**: Ensures that data remains accurate and consistent over its lifecycle by validating key constraints.
+2. **Missing Data**: Identifies and validates gaps in datasets to [maintain completeness](../missingness) and usability.
```
```suggestion
2. **Missing Data**: Identifies and validates gaps in datasets to [maintain completeness](/reference/learn/data_quality_use_cases/missingness.md) and usability.
```

@Quantisan This should fix the Netlify build - Docusaurus pathing works relative to its `docs/` folder. It's building for me locally, let me know if you run into any problems.

Quantisan replied:
thanks, @rachhouse! I missed that error.

Quantisan replied:
can you point me to the script that we can run to build this locally? Better yet, is there a hot reload feature?

rachhouse replied:
Our most current docs for building the docs locally are here: https://github.com/great-expectations/great_expectations/tree/develop/docs/docusaurus

However, it's not the clearest procedure - happy to walk through it on screen during this week's call if that will be helpful.

Quantisan committed:
Update docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md

Co-authored-by: Rachel House <rachel.house@greatexpectations.io>

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,296 @@
+
+Explore core Expectations for schema validation, delving into their practical applications and nuances:
+
+### `expect_column_to_exist`
```
I like the switch to H3 headings to help visually separate the Expectation subsections, but I don't like the way it renders on the right sidebar nav. 
<img width="353" alt="sidebar_nav_with_backticked_h3" src="https://github.com/great-expectations/great_expectations/assets/24724702/b8af7124-5ffd-41ea-9554-f05fd015304f">
Can update the heading text to reflect the Expectation's GX Cloud name in non-backticked text (e.g. **Expect Column to Exist** instead of `expect_column_to_exist`)? Let's see how that looks.

Quantisan replied:
I tried a few variations and settled on spaced caps without quotes. It shows better on both sidebar and body.

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,296 @@
+
+### `expect_column_to_exist`
+
+Ensures the presence of a specified column in your dataset. This Expectation is foundational for
```
I'd like us to also link each Expectation to its Expectation Gallery page (e.g. https://greatexpectations.io/expectations/expect_column_to_exist for `expect_column_to_exist`). 

I'm open to how we include the link - two ideas are:
* Hyperlinked to the "...dataset. [This Expectation](https://greatexpectations.io/expectations/expect_column_to_exist) is foundational..." 
* Added as a distinct item in the section

Quantisan replied:
I opted for an explicit link with clear text anchor as to where it links to.

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,296 @@
+issues before they propagate through your data ecosystem. This guide will walk you through
+leveraging these Expectations to implement effective schema validation in your data workflows.
+
+## Data Preview
```
H2/H3/H4+ headings should be sentenced cased as well, with the exception of the Expectation name H3 headings once switched to the GX Cloud naming convention (see my comment on Expect Column to Exist). For example:
* Data Preview -> Data preview
* Key schema Expectations -> Key schema Expectations
* Examples & Scenarios -> Examples and scenarios
* Common Pitfalls and How to Avoid Them -> Common pitfalls and how to avoid them

Admittedly, it does look a bit awkward for some of the shorter headings with sentence casing and proper nouns, but I think it's best to maintain consistency with the GX docs style.

Quantisan committed:
sentence casing all headings

Quantisan committed:
edit Expectation headings to wrap better on right sidebar

https://github.com/great-expectations/great_expectations/pull/10022#discussion_r1661156027

Quantisan committed:
add links to Expectation Gallery

https://github.com/great-expectations/great_expectations/pull/10022#discussion_r1661185378

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,304 @@
+robust schema validation within your data pipelines, helping to catch and address schema-related
+issues before they propagate through your data ecosystem. This guide will walk you through
+leveraging these Expectations to implement effective schema validation in your data workflows.
+
```
```suggestion
## Prerequisite knowledge

This article assumes basic familiarity with GX terminology, components, and workflows. If you haven't had a chance to get to know GX yet, check out these guides first:
* **GX Cloud**: [GX Cloud workflow](/cloud/about_gx.md#gx-cloud-workflow)
* **GX Core**: [Core components](/core/introduction/gx_overview.md#gx-core-components)
```

As discussed, we want to add a prereqs section before _Data preview_ to make it clear to the user that the article assumes familiarity with GX terminology and the different components of a GX workflow. This is the representative text and links I used when showing what it would look like to Abe - I welcome your wordsmithing and shaping so that it caveats, but doesn't overwhelm the user as they begin reading the article.

I'll take a look through our docs and see if there is an appropriate single or simplified resource that we can provide in this section.

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,304 @@
+
+This dataset includes columns like `sender_account_number`, `recipient_fullname`, `transfer_amount`, and `transfer_date`.
+
+## Key schema Expectations
```
To recap our discussion:
* Let's create two H3 groupings under Key schema Expectations, _Column-level Expectations_ and _Table-level Expectations_. I'm open to the heading wording if you have other suggestions!
* The current Expectations headers move to H4 under their respective parent section (Column-level or Table-level)
* Each Expectation retains its current description, code snippet, and use case.
* Use case format can be changed from the admonition block back to its original text presentation (bold **Use case:** followed by the text)
---

**Use case:** This is the Expectation use case text...

---

* Tips can be consolidated as appropriate at the parent grouping level, each Expectation does not have to have an individual tip. Feel free to play around with where best to present the tips (still in the admonition boxes).

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,304 @@
+Ensures the presence of a specified column in your dataset. This Expectation is foundational for
+schema validation, verifying that critical columns are included, thus preventing data processing
+errors due to missing fields. ([Link to Expectation
+Gallery](https://greatexpectations.io/expectations/expect_column_to_exist))
```
As mentioned, I'm going to see if we can find a less visually distracting mechanism in Docusaurus to provide a link to the Expectation Gallery page for readers who want to check out the details.

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,304 @@
+
+- **Logging and Monitoring**: Even the best validation setup can fail without proper logging and monitoring. Undetected schema validation failures can propagate through the data pipeline unnoticed, leading to broader issues. Detailed logging and real-time monitoring are essential to create an audit trail and enable quick detection and resolution of schema validation problems, maintaining the inte ...
+
+### Additional resources on the web
```
As discussed, let's remove the external links from the documentation page, but feel free to incorporate into the associated blog post!

rachhouse committed:
[DOCS] Add template for code snippet under test for DQ schema use case (#10063)

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

rachhouse committed:
resolve merge conflicts in examples_under_test.py

rachhouse committed:
lint fix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rachhouse committed:
add GX Cloud gifs

Quantisan committed:
remove Additional Resources on the Web section. No external links in tech docs.

Quantisan committed:
update schema code to 1.0 and refactor into CI workflow

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,246 @@
+
+**Insight**: Dataset 2 fails to validate due to the absence of `recipient_fullname` in one of the rows and the correct column count, highlighting how missing critical columns can disrupt financial processing or lead to compliance issues.
+
+### Different Expectation suites: Strict vs. relaxed type checking
```
I think the code recipe in this section could use some improvement, here are my suggestions:
* Validate a single dataset using two different Expectation Suites to highlight the difference in results, instead of validating two datasets with two different Expectation Suites.
* The two separate examples are mostly duplicated code, with the exception of the Expectation Suites, which makes this section difficult to parse as a reader. Can you please consolidate the code into a single workflow (script) that defines the Data Source (using one dataset) and two Expectation Suites that are then run against the same Data Source? I think naming the Suites `strict_suite` and `relaxed_suite` (or something similar - feel free to use what naming you think best) will make it clearer to the user what we're trying to communicate.
* I think it would be beneficial to also include `ExpectColumnValuesToBeOfType` and `ExpectColumnValuesToBeInTypeList` in the example Suites to illustrate strict vs. relaxed validation, respectively.

rachhouse replied:
@ixf-tdg4gx As part of refactoring the organization of the schema code under test so that we contain all the schema-related scripts in a subdirectory, I consolidated the relaxed and strict validation examples into a single script: `docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema_resources/schema_strict_and_relaxed.py`
If you could tackle the third bullet (adding type checking to the strict/relaxed suites in addition to column set), that would be great. Thanks!

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.py:
```diff
@@ -0,0 +1,79 @@
+
+print(f"Validation results 1:\n{results_1}")
+print(f"Validation results 2:\n{results_2}")
+# </snippet>
```
Can you add some CI tests after the snippet to verify that we're getting the expected results from the validation on the sample data? This script will be run with Pytest, so `assert`s are fine.

ixf-tdg4gx replied:
Could you provide an example how you usually do this? 🙇🏻

rachhouse replied:
@ixf-tdg4gx Thanks for asking! I checked in with Engineering on this topic and stand corrected - for code samples/snippets that we include in the docs, all we need to verify is that our scripts containing the sample code run end-to-end without throwing an Exception/error. So, the current approach is fine as-is, no need for extra `asserts`.

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
````diff
@@ -1,8 +1,246 @@
+errors due to missing fields. ([Link to Expectation
+Gallery](https://greatexpectations.io/expectations/expect_column_to_exist))
+
+```python
````
Sorry, forgot to include this comment with my review a moment ago - can you please put the Expectation code snippets under test as well?

ixf-tdg4gx replied:
Sorry, I can't follow you here. What would you like to have under test as well?! 😕

rachhouse replied:
@ixf-tdg4gx As part of refactoring the organization of the schema code under test, I created a new script, `docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema_resources/schema_expectations.py`, which provides an example of how to test the Expectation snippets. Let me know if you have any questions!

To underscore its value, adding this test [exposed an incorrectly formatted snippet](https://github.com/great-expectations/great_expectations/actions/runs/9880224813/job/27288509311?pr=10022#step:11:1747). 😄  I think that ExpectTableColumnsToMatchOrderedList just needs to provide `column_list` as a named argument.

ixf-tdg4gx replied:
Yeah, that was exactly it 👍🏻

rachhouse committed:
reorganize Expectations and gallery links per Abe's feedback

rachhouse committed:
reorganize schema use case code and create tested snippets for Expectations

rachhouse committed:
Merge branch 'develop' into docs/data-quality-schema-tech-doc

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rachhouse committed:
fix broken link

rachhouse committed:
duh

Quantisan committed:
add intro to parent (i.e. column or table) Expectations

Quantisan committed:
add "GX tip for table-level Expectations"

Quantisan committed:
add internal links in the Avoid common pitfalls section

Quantisan committed:
remove individual GX Tip admonitions

Commit:
add missing column_list parameter name to make ExpectTableColumnsToMatchOrderedList expectation pass test

Commit:
add type checking expectations

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rachhouse committed:
small edits prior to marketing and product review, expand DQ use case landing page

rachhouse committed:
fix md typo

rachhouse committed:
Merge branch 'develop' into docs/data-quality-schema-tech-doc

rachhouse committed:
replace first recipe with another that will work with Cloud for foreseeable future

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rachhouse committed:
ok lint

rachhouse committed:
Merge branch 'develop' into docs/data-quality-schema-tech-doc

rachhouse committed:
see if asserts are the issue

rachhouse committed:
re-add asserts

rachhouse committed:
Merge branch 'develop' into docs/data-quality-schema-tech-doc

rachhouse committed:
update psycopg2 connection string to actually use ci test db

rachhouse committed:
Merge branch 'develop' into docs/data-quality-schema-tech-doc

rachhouse committed:
update to emphasize complement of Cloud and Core

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rachhouse committed:
Merge branch 'develop' into docs/data-quality-schema-tech-doc

rachhouse committed:
results, not suite

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rachhouse committed:
tests fixed. hopefully

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rachhouse committed:
Merge branch 'develop' into docs/data-quality-schema-tech-doc

rachhouse committed:
tweak example wording

rachhouse committed:
Merge branch 'develop' into docs/data-quality-schema-tech-doc

joshzzheng commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/schema.md:
```diff
@@ -1,8 +1,235 @@
+
+This dataset is representative of financial transfers recorded by banking institutions. Its fields include account type, sender account, sender name, transfer amount, and transfer date.
+
+You can [access this dataset](https://raw.githubusercontent.com/great-expectations/great_expectations/develop/tests/test_sets/learn_data_quality_use_cases/schema_financial_transfers.csv) from the `great_expectations` GitHub repo in order to reproduce the code recipes provided in this article.
```
Link is 404 right now

rachhouse replied:
Yep, once the PR is merged, then the test data file (new in this PR) will then be available in the `great_expectations` repo at that location.

rachhouse committed:
Merge branch 'develop' into docs/data-quality-schema-tech-doc

rachhouse commented:
@joshzzheng @abegong This content was approved by Hernan & Thaise in internal Slack conversation - just a heads up that I'm approving on the PR so it can be merged in.

rachhouse reviewed (approved):
Approving based on internal approval via Slack.
